[TIMOB-23406] Adding 'scrolling' event to ListView to track scrolling…

rlustemberg opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-23406

At the moment is not possible to determine the scroll direction of a ListView natively and as the view is scrolling. Very common effects such as the hiding of the navigation bar when scrolling (like on the Facebook app) become impossible to achieve without cumbersome workarounds.
I will send a PR with a modification to the SDK core.

rlustemberg committed:
[TIMOB-23406] Adding 'scrolling' event to ListView to track scrolling direction

nuno commented:
+1

hansemannn commented on iphone/Classes/TiUIListView.m:
```diff
@@ -1815,7 +1815,24 @@ - (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
+    if ([self.proxy _hasListeners:@"scrolling"]) {
+        NSMutableDictionary *eventArgs = [NSMutableDictionary dictionary];
+        NSString* swipeString = @"none";
+        if (velocity.y > 0){
```
Missing space in `) {`.

hansemannn commented on iphone/Classes/TiUIListView.m:
```diff
@@ -1815,7 +1815,24 @@ - (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
+        if (velocity.y > 0){
+            swipeString = @"up";
+        }
+        if (velocity.y < 0){
```
Missing space in `) {`.

hansemannn commented on iphone/Classes/TiUIListView.m:
```diff
@@ -1815,7 +1815,24 @@ - (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
+        if (velocity.y < 0){
+            swipeString = @"down";
+        }
+        if(swipeString !=(id)[NSNull null])
```
It can never be null, because you set the default value to `none`. So please leave the default value to nil and only check `if (swipeString != nil) {`. Also note that the bracket goes into the same line.

rlustemberg replied:
Yes, I'll take it out. It's a leftover from when I was not setting a default (none)

rlustemberg replied:
If I initialise the swipeString to nil and do a nil check before firing the event, eventually I get a bad access exception. I left if initialised as 'none' and I do if (swipeString != @"none") instead. In that way it doesn't crash.

hansemannn commented:
Left some comments. Also please add docs for the new event.

hansemannn commented on iphone/Classes/TiUIListView.m:
```diff
@@ -1815,7 +1815,24 @@ - (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
     }
 }
-
+-(void)scrollViewWillEndDragging:(UIScrollView *)scrollView withVelocity:(CGPoint)velocity targetContentOffset:(inout CGPoint *)targetContentOffset{
```
Think about including the velocity and contentOffset as well, so the developer can receive as much information as possible.

rlustemberg replied:
Are the comments for me or for you guys? ;)

hansemannn replied:
For you, since you are the contributor and we are the reviewer 😊

rlustemberg replied:
LOL.
Will do my best to get this ready before Monday. Greetings from Amsterdam, Hans!

rlustemberg committed:
Added velocity and targetContentOffset to the event payload

rlustemberg committed:
[TIMOB-23406] Added scrolling ListView event to docs

rlustemberg commented:
I've added velocity (y axis) and the full x,y coordinates for targetContentOffset as I couldn't make up my mind of a name for only the y point (x is always 0).
I've renamed swipeString to directionString and kept it initialised because for some reason unknown to me , initialised as nil and doing a nil check before firing the event will still through an bad access exception.

rlustemberg committed:
[TIMOB-23406]  Adding spaces

rlustemberg committed:
[TIMOB-23406] Fixed broken yaml

hansemannn commented on apidoc/Titanium/UI/ListView.yml:
```diff
@@ -598,6 +598,31 @@ events:
+        summary: The expected offset when the scrolling action decelerates to a stop.
+        type: Object
+
+      - name: bubbles
```
If it doesn't bubble, remove the property.

hansemannn commented on apidoc/Titanium/UI/ListView.yml:
```diff
@@ -598,6 +598,31 @@ events:
+        type: number
+
+      - name: targetContentOffset
+        summary: The expected offset when the scrolling action decelerates to a stop.
```
Mention that the object consists of an `x` and `y` key.

hansemannn commented on iphone/Classes/TiUIListView.m:
```diff
@@ -1815,7 +1815,32 @@ - (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
+            directionString = @"down";
+        }
+        
+        if(directionString != @"none") {
```
This is bad code-style, we need to do a proper-nil check here.

rlustemberg replied:
I know, but just checking for nil was still not working. Please remember I'm no Objective C dev

hansemannn commented on iphone/Classes/TiUIListView.m:
```diff
@@ -1815,7 +1815,32 @@ - (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
+            NSMutableDictionary *eventArgs = [NSMutableDictionary dictionary];
+            NSDictionary *pointObject = [NSDictionary dictionaryWithObjectsAndKeys:
+             [NSNumber numberWithDouble:targetContentOffset->x],@"x",
+             [NSNumber numberWithDouble:targetContentOffset->y],@"y",
```
Don't access the properties like this. I will provide a gist on a proper usage.

hansemannn commented on iphone/Classes/TiUIListView.m:
```diff
@@ -1815,7 +1815,32 @@ - (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
+            [eventArgs setValue:directionString forKey:@"direction"];
+            [eventArgs setValue:NUMDOUBLE(velocity.y) forKey:@"velocity"];
+            [eventArgs setObject:pointObject forKey:@"targetContentOffset"];
+            [self.proxy fireEvent:@"scrolling" withObject:eventArgs withSource:self.proxy propagate:NO reportSuccess:NO errorCode:0 message:nil];
```
You need to check if the event-listener is added beforehand. Will also mention it in my gist.

rlustemberg committed:
[TIMOB-23406] updated doc for 'scrolling' event

hansemannn commented:
Here is what I would do: https://gist.github.com/hansemannn/604ad5060bd69710d0326580ebf9ac90

Also what I noticed:
- The direction is always then false, when the user drags without releasing the finger. Also the velocity is0 then.
- When the direction is `down`, the velocity is negative.
- The `targetContentOffset` key `x` is always 0 (of course, since we can only scroll vertically).

That should also be added to the docs to clarify. Please test the gist and update your code and docs if it makes sense to you.

rlustemberg committed:
[TIMOB-23406] Added corrections from @hansemannn

hansemannn commented on iphone/Classes/TiUIListView.m:
```diff
@@ -1815,7 +1815,35 @@ - (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
+                                                                                             @"y": NUMFLOAT(targetContentOffset->y),
+                                                                                             },
+                                                                                     @"velocity": NUMFLOAT(velocity.y)
+                                                                                     }];
```
The copy-and-paste caused formatting issues here, please refer to the gist and format manually.

hansemannn commented on apidoc/Titanium/UI/ListView.yml:
```diff
@@ -598,6 +598,27 @@ events:
+    since: 5.4.0
+    properties:
+      - name: direction
+        summary: Direction of the scroll--either 'up', or 'down'.
```
Change to: "Direction of the scroll, either `up`, or `down`."

hansemannn commented on apidoc/Titanium/UI/ListView.yml:
```diff
@@ -598,6 +598,27 @@ events:
+        type: number
+
+      - name: targetContentOffset
+        summary: The expected offset as an object consisting of x and y keys when the scrolling action decelerates to a stop. 
```
Break this line to
```yml
summary: |
    The expected offset as an object consisting of x and y keys when 
    the scrolling action decelerates to a stop. 
```

rlustemberg committed:
Updated docs and formatting

rlustemberg committed:
[TIMOB-23406] added lines

rlustemberg committed:
Modified event object to just send the y axis value of targetContentOffset and updated docs accordingly

hansemannn commented on apidoc/Titanium/UI/ListView.yml:
```diff
@@ -598,6 +598,31 @@ events:
+
+      - name: velocity
+        summary: The velocity of the scroll in scale factor per second
+        type: number
```
This has to be `Number`, otherwise the doc validation fails.

hansemannn commented on apidoc/Titanium/UI/ListView.yml:
```diff
@@ -598,6 +598,31 @@ events:
+      - name: targetContentOffset
+        summary: |
+           The expected y axis offset when the scrolling action decelerates to a stop. 
+        type: number
```
Same here: This has to be `Number`, otherwise the doc validation fails.

rlustemberg committed:
Update ListView.yml

hansemannn commented:
Approved!
